File: scala/Integration.scala
package org.github.antma.cpalgo
object Integration {
  def simpson(a: Double, b: Double, eps: Double, f : Double => Double): Double = {
    val fa = f(a)
    val fb = f(b)
    val avg = (fa + fb) * 0.5
    val l = (b - a)
    val l3 = l / 3.0
    def calc(n: Int, s1: Double, s2: Double): Double = {
      (l3 / n) * (avg + s1 + 2.0 * s2)
    }
    def loop(n: Int, s1: Double, pi: Double): Double = {
      val step = l / n
      val s2 = (0 until n).map {k => f(a + (k + 0.5) * step) }.sum 
      val i = calc(n, s1, s2)
      if((i - pi).abs < eps) i
      else loop(2 * n, s1 + s2, i)
    }
    loop(2, f((a + b) * 0.5), 0.0)
  }
}

File: rust/src/integration.rs
pub fn simpson<F>(a: f64, b: f64, eps: f64, f: F) -> f64
where
  F: Fn(f64) -> f64,
{
  let fa = f(a);
  let fb = f(b);
  let avg = (fa + fb) * 0.5;
  let l = b - a;
  let l3 = l / 3.0;
  let mut n = 2;
  let mut s1 = f((a + b) * 0.5);
  let mut pi = 0.0;
  loop {
    let invn = 1.0 / (n as f64);
    let step = l * invn;
    let mut s2 = 0.0;
    let mut x = a + step * 0.5;
    for _ in 0..n {
      s2 += f(x);
      x += step;
    }
    let i = l3 * invn * (avg + s1 + 2.0 * s2);
    if (i - pi).abs() < eps {
      break i;
    }
    n *= 2;
    s1 += s2;
    pi = i;
  }
}
